aws.go

```diff
@@ -0,0 +1,439 @@
+	svc       *dynamodb.DynamoDB // Reference to the creators table
+	common
+}
+
+// NodeItem is the dynamodb table item representation of a node
+type NodeItem struct {
+	Network      string    // The name of the network the node belongs to
+	Domain       string    // The domain name associated with the node
+	Created      time.Time // The time that the node first came online
+	Expires      time.Time // The time that the node will retire from the network
+	Role         int       // The role the node has in the network
+	ScramblerKey string    // Secret used to scramble data with fixed nonce
+}
+
+// SecretItem is the dynamodb table item representation of a secret
+type SecretItem struct {
+	Domain       string
+	TimeStamp    time.Time
+	ScramblerKey string
+}
+
+// NewAWS creates a new instance of the AWS structure
+func NewAWS(region string) (*AWS, error) {
+	var a AWS
+	var sess *session.Session
+
+	// Configure session with credentials from .aws/credentials and region...
+	if region != "" {
+		// ... from region param
+		config := &aws.Config{
+			Region: aws.String(region),
+		}
+		sess = session.Must(session.NewSession(config))
+	} else {
+		// ... from .aws/config
+		sess = session.Must(session.NewSessionWithOptions(session.Options{
+			SharedConfigState: session.SharedConfigEnable,
+		}))
+	}
+	if sess == nil {
+		return nil, errors.New("AWS session is nil")
+	}
+	a.svc = dynamodb.New(sess)
+
+	_, err := a.awsCreateTables()
+	if err != nil {
+		return nil, err
+	}
+
+	a.mutex = &sync.Mutex{}
+	err = a.refresh()
+	if err != nil {
+		return nil, err
+	}
+	return &a, nil
+}
+
+func (a *AWS) awsCreateTables() (bool, error) {
+
+	_, err := a.createNodesTable()
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case dynamodb.ErrCodeTableAlreadyExistsException:
+				break
+			case dynamodb.ErrCodeResourceInUseException:
+				break
+			default:
+				return false, err
+			}
+		} else {
+			return false, err
+		}
+	}
+
+	_, err = a.createSecretsTable()
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case dynamodb.ErrCodeTableAlreadyExistsException:
+				break
+			case dynamodb.ErrCodeResourceInUseException:
+				break
+			default:
+				return false, err
+			}
+		} else {
+			return false, err
+		}
+	}
+
+	for {
+		input := &dynamodb.DescribeTableInput{
+			TableName: aws.String(nodesTableName),
+		}
+		result, err := a.svc.DescribeTable(input)
+		if err != nil {
+			return false, err
+		}
+		if *result.Table.TableStatus == "ACTIVE" {
+			break
+		}
+	}
+
+	for {
+		input := &dynamodb.DescribeTableInput{
+			TableName: aws.String(secretsTableName),
+		}
+		result, err := a.svc.DescribeTable(input)
+		if err != nil {
+			return false, err
+		}
+		if *result.Table.TableStatus == "ACTIVE" {
+			break
+		}
+	}
+
+	return true, nil
+}
+
+func (a *AWS) createNodesTable() (*dynamodb.CreateTableOutput, error) {
```

Comment:
Not sure about the business logic here, but shouldn't we add the TTL on the expires column? Otherwise the records in this table will only grow, the scan will take ages and eventually fill up whatever memory is out there.
Or am i wrong?

Reply:
That makes sense, looks straightforward to do for nodes but will need to modify the secrets item struct to include expiry

Reply:
Will make this a separate PR for this.